content/en/blog/_posts/2022-09-14-pod-has-network-condition.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: blog
+title: 'Kubernetes 1.25: PodHasNetwork condition for pods'
+date: 2022-09-14
+slug: pod-has-network-condition
+---
+
+**Author:**
+Deep Debroy (Apple)
+
+Kubernetes 1.25 introduces Alpha support for a new kubelet-managed pod condition
+in the status field of a pod: `PodHasNetwork`. The kubelet will use the
+`PodHasNetwork` condition to accurately surface the initialization state of a
+pod from the perspective of pod sandbox creation and network configuration by a
+CRI plugin in coordination with CNI plugins. The kubelet starts to pull
+container images and start individual containers (including init containers)
```

Comment:
That's technically a bit incorrect.
Kubelet is only asking CRI to do image(s) pulling, then it is asking CRI to prepare the sandbox (that's the time when CNI plugin will be called - btw. plugIN, not plugINS - there is call only to a single plugin, which in fact can be a CNI multiplexer, like Multus or DANM, but there is a call to only a single plugin in all of known to me CRI implementations), and then - to run containers using these images in that particular sandbox.

Reply:
> Kubelet is only asking CRI to do image(s) pulling, then it is asking CRI to prepare the sandbox

To clarify, the sequence is: 
First, Kubelet asks the runtime to prepare the sandbox (including network config through CNI) [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L664].

Once sandbox creation succeeds, Kubelet asks the runtime to pull images for individual containers in the pod and start/run the containers [https://github.com/kubernetes/kubernetes/blob/d0e413e86d150e0d4638ce306ca3fd437fe0b763/pkg/kubelet/kuberuntime/kuberuntime_container.go#L173]

Pulling of container images does not take place before the sandbox creation has successfully concluded.

Reply:
@ddebroy big plus for more precise description!

Reply:
> btw. plugIN, not plugINS

The CNI spec specifies chaining while allows multiple plugins to be specified and invoked in sequence right? https://github.com/containernetworking/cni/blob/eeded9f55ef7a23ee3b1f6cadbc8d51f879238a9/SPEC.md#network-configuration-lists

Reply:
yes. you are right and my comment was based on simplification (common case) rather than wider scenario.
also cni lib which is the default way of handling that can do that list chaining, but it is not related to multinetwork (or even multi topic, but rather to tune after the tune after the tune of something what was produced by first in the chain plugin.
multi network topic is still "out of the tree discussion" which was blocked for several years because it was not needed by google, but... looks like even in google several people are working on that :]
anyway - right now for having multiple networks (not multi address types) in kubernetes you need some kind of multiplexer (most commonly used is Multus, but there are other implementations of that, which are using the common "out of kubernetes" CRD standard - https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd

Reply:
CNI is a choice of the CRI implementation, it doesn't have to use CNI, it can use other techniques to create the network

Reply:
Clarified this. For reference, https://github.com/maiqueb/multus-dynamic-networks-controller appears to be a concrete implementation of what was being referred to above.

Reply:
Clarified with `(typically in coordination with CNI plugins)`